lib/access.js

```diff
@@ -34,152 +20,200 @@ const subcommands = [
   '2fa-not-required',
 ]
 
-const UsageError = (msg) =>
-  Object.assign(new Error(`\nUsage: ${msg}\n\n` + usage), {
-    code: 'EUSAGE',
-  })
-
-const cmd = (args, cb) =>
-  access(args)
-    .then(x => cb(null, x))
-    .catch(err => err.code === 'EUSAGE'
-      ? cb(err.message)
-      : cb(err)
+class Access {
+  constructor (npm) {
+    this.npm = npm
+  }
+
+  // TODO we don't use this, why are we exporting it?
+  get subcommands () {
+    return subcommands
+  }
+
+  get usage () {
+    return usageUtil(
+      'npm access',
+      'npm access public [<package>]\n' +
+      'npm access restricted [<package>]\n' +
+      'npm access grant <read-only|read-write> <scope:team> [<package>]\n' +
+      'npm access revoke <scope:team> [<package>]\n' +
+      'npm access 2fa-required [<package>]\n' +
+      'npm access 2fa-not-required [<package>]\n' +
+      'npm access ls-packages [<user>|<scope>|<scope:team>]\n' +
+      'npm access ls-collaborators [<package> [<user>]]\n' +
+      'npm access edit [<package>]'
     )
+  }
 
-const access = async ([cmd, ...args], cb) => {
-  const fn = subcommands.includes(cmd) && access[cmd]
+  async completion (opts) {
+    const argv = opts.conf.argv.remain
+    if (argv.length === 2)
+      return subcommands
+
+    switch (argv[2]) {
+      case 'grant':
+        if (argv.length === 3)
+          return ['read-only', 'read-write']
+        else
+          return []
+
+      case 'public':
+      case 'restricted':
+      case 'ls-packages':
+      case 'ls-collaborators':
+      case 'edit':
+      case '2fa-required':
+      case '2fa-not-required':
+      case 'revoke':
+        return []
+      default:
+        throw new Error(argv[2] + ' not recognized')
+    }
+  }
 
-  if (!cmd)
-    throw UsageError('Subcommand is required.')
+  exec (args, cb) {
```

Comment:
It looks like you're heading in this direction already, but pretty please can we ditch the callbacks entirely? 😆  I'd love to see `this.access` just be `this.exec()`, and have npm know how to handle that.

Reply:
That change was left out of this PR because of the amount of testing changes it would require.  This PR is mostly code change, with minimal test changes.

Making this function async would be a minimal code change, and a larger test change.  It'd be much easier than THIS PR, but still needs to be its own thing.

Reply:
That test already exists in `test/lib/load-all-commands.js`